src/serde/deserialize_utils.rs

```diff
@@ -229,135 +196,130 @@ mod tests {
-        };
+        assert_eq!(parsed, Ok((" + (-1)", Register::FP)));
+    }
 
-        assert_eq!(value_address, parsed_value);
+    #[test]
+    fn parse_offset_test() {
+        let value_1 = " + (-1)";
+        let parsed_1 = offset(value_1);
+        assert_eq!(parsed_1, Ok(("", -1_i32)));
+
+        let value_2 = " + 1";
+        let parsed_2 = offset(value_2);
+        assert_eq!(parsed_2, Ok(("", 1_i32)));
     }
 
     #[test]
-    fn parse_reference_with_one_offset_test() {
-        let value_string: &str = "cast(fp + (-3), felt*)";
-        let splitted_value: Vec<&str> = value_string.split(" + ").collect();
+    fn parse_register_and_offset_test() {
+        let value_1 = "fp + 1";
+        let parsed_1 = register_and_offset(value_1);
 
-        let parsed_value = parse_reference_with_one_offset(&splitted_value).unwrap();
+        assert_eq!(parsed_1, Ok(("", (Register::FP, 1_i32))));
 
-        let value_address = ValueAddress {
-            register: Some(Register::FP),
-            offset1: -3,
-            offset2: 0,
-            immediate: None,
-            dereference: false,
-            inner_dereference: false,
-        };
+        let value_2 = "ap + (-1)";
+        let parsed_2 = register_and_offset(value_2);
 
-        assert_eq!(value_address, parsed_value);
+        assert_eq!(parsed_2, Ok(("", (Register::AP, -1_i32))));
     }
 
     #[test]
-    fn parse_reference_with_two_offsets_test() {
-        let value_string: &str = "cast([fp + (-4)] + 1, felt*)";
-        let splitted_value: Vec<&str> = value_string.split(" + ").collect();
+    fn parse_inner_dereference_test() {
+        let value = "[fp + (-1)] + 2";
+        let parsed = inner_dereference(value);
 
-        let parsed_value = parse_reference_with_two_offsets(&splitted_value).unwrap();
-
-        let value_address = ValueAddress {
-            register: Some(Register::FP),
-            offset1: -4,
-            offset2: 0,
-            immediate: Some(bigint!(1)),
-            dereference: false,
-            inner_dereference: false,
-        };
-
-        assert_eq!(value_address, parsed_value);
+        assert_eq!(parsed, Ok((" + 2", (true, Register::FP, -1_i32))));
     }
 
     #[test]
-    fn parse_reference_no_offsets_test() {
-        let value_string: &str = "cast(fp, felt*)";
-        let splitted_value: Vec<&str> = value_string.split(" + ").collect();
+    fn parse_no_inner_dereference_test() {
+        let value = "ap + 3";
+        let parsed = no_inner_dereference(value);
 
-        let parsed_value = parse_reference_no_offsets(&splitted_value).unwrap();
-
-        let value_address = ValueAddress {
-            register: Some(Register::FP),
-            offset1: 0,
-            offset2: 0,
-            immediate: None,
-            dereference: false,
-            inner_dereference: false,
-        };
+        assert_eq!(parsed, Ok(("", (false, Register::AP, 3_i32))));
+    }
 
-        assert_eq!(value_address, parsed_value);
+    #[test]
```

Comment:
These tests are good but since they are completely independent I would have them in three separate test functions.